### PR TITLE
feat: notification worker when the user loses streak

### PR DESCRIPTION
**Pull request by sshanzel**
NOTE: this should not be merged until we have the feature to restore a streak, otherwise, we will be sending notifications already without having the ability to actually restore the streak.

- When the user's streak gets updated and the current streak becomes 0 from a non-zero value, we should set the cache.
- We will use the cache's value to identify how much was the lost streak.

### CODE REVIEW

File: src/workers/cdc/primary.ts [hunk 1]
```diff
@@ -1,85 +1,84 @@
 import {
-  UserState,
```
Comment (sshanzel) [inline src/workers/cdc/primary.ts:2]: Lint annoyingly rearranged the imports on its own 😅

File: src/workers/cdc/primary.ts [hunk 2]
```diff
@@ -795,12 +795,42 @@ const onSquadPublicRequestChange = async (
...
 
+const setRestoreStreakCache = async (
+  userId: string,
+  lastViewAt: Date,
+  previousStreak: number,
+) => {
+  const today = new Date();
+  const lastViewDifference = differenceInDays(today, lastViewAt);
+
+  if (lastViewDifference !== 2) {
```
Comment (sshanzel) [inline src/workers/cdc/primary.ts:806]: - It will never be less than 2 as we don't restore when it has only been a day since user can still read something for the streak today.
- When it is more than 2 days now, it means the user is not eligible anymore to restore it. We only allow it on the day you lose the streak (until the day ends).

**Comment by sshanzel** [comment]
@dailydotdev/web-team feel free to send a review. I just thinking about more tests, but the flow should be as what is intended.

**Review by rebelchris — APPROVED** [review]

File: src/entity/Alerts.ts [hunk 3]
```diff
@@ -44,6 +44,9 @@ export class Alerts {
   @Column({ type: 'bool', default: false })
   showStreakMilestone: boolean;
 
+  @Column({ type: 'bool', default: false })
+  showResetStreak: boolean;
```
Comment (capJavert) [inline src/entity/Alerts.ts:48]: If we are saving to redis do we need this? Feels like we would need one or the other no point in saving on multiple places?

File: src/common/utils.ts [hunk 4]
```diff
@@ -110,3 +110,7 @@ export const safeJSONParse = <T>(json: string): T | undefined => {
     return undefined;
   }
 };
+
+export function isNumber(value: string | number): boolean {
+  return !!value && !isNaN(Number(value.toString()));
```
Comment (capJavert) [inline src/common/utils.ts:115]: this would fail if passed `0`

Comment (capJavert) [reply src/common/utils.ts:115]: we have `lodash.isnumber` which should be more bulletproof, we can add it to dependencies

File: src/notifications/generate.ts [hunk 5]
```diff
@@ -143,6 +146,15 @@ export const generateNotificationMap: Record<
       .avatarSource(ctx.source)
       .uniqueKey(ctx.bookmark.remindAt.toString())
       .objectPost(ctx.post, ctx.source, ctx.sharedPost),
+  streak_reset_restore: (builder, ctx: NotificationStreakContext) =>
+    builder
+      .icon(NotificationIcon.Streak)
+      .description('Click here if you wish to restore your streak')
+      .uniqueKey(ctx.streak.lastViewAt.toString())
```
Comment (capJavert) [inline src/notifications/generate.ts:153]: could we make it per day (without time portion) eg `19-08-2024` so we don't duplicate notifications for any reason in the same day?

File: src/common/users.ts [hunk 6]
```diff
@@ -431,6 +444,32 @@ export const shouldResetStreak = (
...
+    if (day === lastDayOfWeek) {
+      return FREEZE_DAYS_IN_A_WEEK;
+    }
+
+    if (day === firstDayOfWeek) {
+      return FREEZE_DAYS_IN_A_WEEK + STREAK_RECOVERY_MAX_GAP_DAYS;
+    }
+
+    return STREAK_RECOVERY_MAX_GAP_DAYS;
+  };
```
Comment (capJavert) [inline src/common/users.ts:468]: lets move function to outside so its not inline initiated each time

File: src/common/users.ts [hunk 7]
```diff
@@ -431,6 +444,32 @@ export const shouldResetStreak = (
...
+
+  return difference - getAllowedDays() === 0;
```
Comment (capJavert) [inline src/common/users.ts:470]: this can technically go to negative, just check if thats fine?

File: src/entity/Alerts.ts (hunk 3 above)
Comment (capJavert) [reply src/entity/Alerts.ts:48]: Might be missing some context but just feels wrong to have two sources of truth.

File: src/common/users.ts (hunk 7 above)
Comment (sshanzel) [reply src/common/users.ts:470]: Yes, we need exact 0 values.

File: src/common/utils.ts (hunk 4 above)
Comment (sshanzel) [reply src/common/utils.ts:115]: Oh, cool. Let me check that.

File: src/entity/Alerts.ts (hunk 3 above)
Comment (ilasw) [inline src/entity/Alerts.ts:48]: Maybe `showRecoverStreak` would be more explicit, what do you think?

Comment (sshanzel) [reply src/entity/Alerts.ts:48]: Good catch! Also that reset is actually a different thing that happens in our system too.

File: src/notifications/generate.ts (hunk 5 above)
Comment (sshanzel) [reply src/notifications/generate.ts:153]: Yeah, was thinking of that too at first, just realized it is unlikely to happen since reset can only trigger after some days, but I agree, we can do that to ensure duplicates will be avoided.

File: src/entity/Alerts.ts (hunk 3 above)
Comment (sshanzel) [reply src/entity/Alerts.ts]: Yeah, at first, I thought this wasn't needed too. But the cache is just to hold the value, which can be viewed from the popup many times until it expires.

This one should hold the value of whether the popup should be shown after boot. Then should not be shown again once seen.

File: src/common/users.ts (hunk 6 above)
Comment (sshanzel) [reply src/common/users.ts]: Agreed. Should be moved by now.

File: src/notifications/generate.ts [hunk 8]
```diff
@@ -143,6 +146,15 @@ export const generateNotificationMap: Record<
       .avatarSource(ctx.source)
       .uniqueKey(ctx.bookmark.remindAt.toString())
       .objectPost(ctx.post, ctx.source, ctx.sharedPost),
+  streak_reset_restore: (builder, ctx: NotificationStreakContext) =>
+    builder
+      .icon(NotificationIcon.Streak)
+      .description('Click here if you wish to restore your streak')
+      .uniqueKey(new Date(ctx.streak.lastViewAt).toISOString().split('T')[0])
```
Comment (capJavert) [inline src/notifications/generate.ts:153]: let's use `format` from date-fns `dd-MM-YYYY`

**Review by capJavert — APPROVED** [review]

File: src/notifications/generate.ts (hunk 8 above)
Comment (sshanzel) [reply src/notifications/generate.ts]: Done 🥳 

Just used `yyyy` - console was throwing an error saying I should use it over the uppercase version.

**Review by nensidosari — APPROVED** [review]

**Review by nensidosari — APPROVED** [review]
Looks good on my side

File: src/workers/cdc/primary.ts [hunk 9]
```diff
@@ -809,12 +817,59 @@ const onSquadPublicRequestChange = async (
...
+
 const onUserStreakChange = async (
   con: DataSource,
   logger: FastifyBaseLogger,
   data: ChangeMessage<UserStreak>,
 ) => {
   if (data.payload.op === 'u') {
+    if (
+      data.payload.after.currentStreak === 0 &&
+      data.payload.before.currentStreak > 2
```
Comment (capJavert) [inline src/workers/cdc/primary.ts:868]: Can we move `2` to some constant, does the day gap one matches?

File: src/entity/user/UserStreakAction.ts [hunk 10]
```diff
@@ -18,7 +18,7 @@ export class UserStreakAction {
   @PrimaryColumn({ type: 'text' })
   type: UserStreakActionType;
 
-  @PrimaryColumn({ type: 'timestamptz', default: () => 'now()' })
+  @PrimaryColumn({ type: 'timestamp', default: () => 'now()' })
```
Comment (rebelchris) [inline src/entity/user/UserStreakAction.ts:21]: More of a request if we can safely modify this and it will not change stored DB values.
Also can we do it for all other `timestamptz` fields?

File: src/notifications/generate.ts [hunk 11]
```diff
@@ -143,6 +147,16 @@ export const generateNotificationMap: Record<
...
+  streak_reset_restore: (builder, ctx: NotificationStreakContext) =>
+    builder
+      .icon(NotificationIcon.Streak)
+      .description('Click here if you wish to restore your streak')
+      .uniqueKey(format(ctx.streak.lastViewAt, 'dd-MM-yyyy'))
+      .targetUrl(notificationsLink)
+      .referenceStreak(ctx.streak)
+      .setTargetUrlParameter([
+        ['streak_restore', ctx.streak.currentStreak.toString()],
+      ]),
```
Comment (capJavert) [inline src/notifications/generate.ts:159]: `streak_restore` is something that triggers popup on the frontend on notifications page?

**Review by rebelchris — COMMENTED** [review]
Overall good, just safekeeping on the DB before me merge please, and request if we can do all in one go rather?

File: src/migration/1723670727196-UserStreakAction.ts [hunk 12]
```diff
@@ -5,7 +5,7 @@ export class UserStreakAction1723670727196 implements MigrationInterface {
 
   public async up(queryRunner: QueryRunner): Promise<void> {
     await queryRunner.query(
-      `CREATE TABLE "user_streak_action" ("userId" character varying(36) NOT NULL, "type" text NOT NULL, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(), CONSTRAINT "PK_875298ad33fc69f7d704042cdd4" PRIMARY KEY ("userId", "type", "createdAt"))`,
```
Comment (capJavert) [inline src/migration/1723670727196-UserStreakAction.ts:8]: Are we gonna manually update this in the DB? ..since this change in migration won't be run on deployment

**Review by capJavert — COMMENTED** [review]
Few sanity checks, other then that good to go!

File: src/entity/user/UserStreakAction.ts (hunk 10 above)
Comment (sshanzel) [reply src/entity/user/UserStreakAction.ts:21]: There is currently no record for this table, so in this instance, we are safe.

File: src/common/utils.ts (hunk 4 above)
Comment (sshanzel) [reply src/common/utils.ts:115]: Ah, yes, forgot to mention. I actually tried it but the test is breaking: https://github.com/dailydotdev/daily-api/pull/2151/commits/a79cdcf82a40fc5828300faf29aa1e859666b010

I think it is due to using `isNaN` vs `Number.isNaN`. The package is probably using the latter, but what we need is the former.

File: src/notifications/generate.ts (hunk 11 above)
Comment (ilasw) [reply src/notifications/generate.ts:159]: Yes, it does!

File: src/workers/cdc/primary.ts (hunk 9 above)
Comment (sshanzel) [reply src/workers/cdc/primary.ts:868]: I think this should be something else, let me update it.

File: src/common/utils.ts (hunk 4 above)
Comment (capJavert) [reply src/common/utils.ts:115]: Ok but implementation above is not correct, if we pass `isNumber(0)` it will say `false` while `0` is a number.

File: src/workers/newView.ts [hunk 13]
```diff
@@ -82,8 +84,24 @@ const incrementReadingStreak = async (
...
+        .getRepository(Alerts)
+        .update({ userId }, { showRecoverStreak: false });
+
+      const key = generateStorageKey(
+        StorageTopic.Streak,
+        StorageKey.Reset,
+        userId,
+      );
+
+      await deleteRedisKey(key);
```
[the remark this answers is not in the document]
Comment (ilasw) [reply src/workers/newView.ts:102]: Sure!

File: src/migration/1723670727196-UserStreakAction.ts (hunk 12 above)
Comment (sshanzel) [reply src/migration/1723670727196-UserStreakAction.ts:8]: Yeah, this is wrong. I initially thought the migration had not yet run in prod. Let me create a new migration.

File: src/entity/user/UserStreakAction.ts (hunk 10 above)
Comment (sshanzel) [reply src/entity/user/UserStreakAction.ts:21]: For the other fields, we should maybe create a separate ticket, so further investigation can be done and the change must be careful since it involves data changes. Wdyt?

Comment (rebelchris) [reply src/entity/user/UserStreakAction.ts:21]: Yeah wasn't aware this was just a new table :)  (empty)
so yeah that makes total sense then.

File: src/common/utils.ts (hunk 4 above)
Comment (capJavert) [reply src/common/utils.ts:115]: It is probably related by the fact you do `!!value` which makes your `isNumber` invalid and you are missing a null check.

File: src/common/users.ts [hunk 14]
```diff
@@ -401,6 +416,12 @@ export const Weekends = [Day.Sunday, Day.Saturday];
 export const FREEZE_DAYS_IN_A_WEEK = Weekends.length;
 export const MISSED_LIMIT = 1;
 
+/*
+ * if last streak was Monday, and today is Wednesday, we should allow recovering streak
+ * if last streak was Friday, then the gap is 3 days (the 24-hour period had passed), we should not allow it
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply src/common/users.ts:421]: Ah ok, got confused by description then.
If it's the case all is good :)

[the remark this answers is not in the document]
Comment (sshanzel) [reply src/common/users.ts:421]: Yeah, thankfully we have tests to prove the cases. The headaches from writing them were worth it 🤣 😭

File: src/common/utils.ts (hunk 4 above)
Comment (capJavert) [reply src/common/utils.ts]: @sshanzel I added some edge case covers based on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN#examples and tests to only support number and string types, see 5fcb1255e38a525d4e3e1a0905f1fffeba3d7c35

**Review by capJavert — APPROVED** [review]
